### ISSUE TITLE
how to process Robust04

### ISSUE THREAD

**Issue by yuhaia** [allenai/ir_datasets#52]
Hi, it's very nice of you guys to build this package to provides a common interface to many IR ranking datasets. And I wanna ask how to process the documents of Robust04 when I have downloaded the raw data after I received the agreements with NIST.

I saw this information when I run the interface of ir_datasets:

> [INFO] The TREC Robust document collection is from TREC disks 4 and 5. Due to the copyrighted nature of the  documents, this collection is for research use only, which requires agreements to be  filed with NIST. See details here: <https://trec.nist.gov/data/cd45/index.html>.
Once completed, place the uncompressed source here: /Users/yuhai09/.ir_datasets/trec-robust04/trec45

I don't know how to process the documents of Robust04 so that it can meet the requirement of ir_datasets. I would appreciate it if you could provide details about it. Thanks a lot!

**Comment by seanmacavaney** [comment]
Hi yuhaia,

You can extract the source files, which should give you a directory (let's call it `/path/to/robust/source`) with two sub-directories: `NEWS_data` and `NEWS_dtd`. Any of the following will then work to give it to ir_datasets:

 - `mv /path/to/robust/source /Users/yuhai09/.ir_datasets/trec-robust04/trec45` (if you want to move the source files)
 - `cp -r /path/to/robust/source /Users/yuhai09/.ir_datasets/trec-robust04/trec45` (if you want to make a copy of the source files)
 - `ln -s /path/to/robust/source /Users/yuhai09/.ir_datasets/trec-robust04/trec45` (if you want to make a symbolic link to the files)

Please let us know if there's a way we could improve the instructions/documentation/user experience for this!

 \- sean

**Comment by seanmacavaney** [comment]
@yuhaia did this work?

**Comment by yuhaia** [comment]
> @yuhaia I recently found out that the way they provide the data has changed since the version I had. This should be fixed with the version on the master branch now. As long as the path `~/.ir_datasets/trec-robust04/trec45` contains `FBIS` `FR94` `FT` `LATIMES` somewhere under it, it should work now (either with compressed or uncompressed document files). Can you give this a try and let me know if it works for you?
> 
> Alternatively, since you have permission to the dataset, you could use this extension: https://github.com/seanmacavaney/irds_robust_anserini, which will download a version of the documents that are publicly available online.

Sorry it's late to reply your help. It's true that the way they provide the Robust04 data has changed and the raw data contains `TREC-Disk-4/` and `TREC-Disk-5` folder. I can see  `FBIS` `FR94` `FT` `LATIMES` in these two folder. But it doesn't work after I put these four folder to  `~/.ir_datasets/trec-robust04/trec45`. When I run follow code 
`import ir_datasets
dataset = ir_datasets.load('trec-robust04/fold1')
for doc in dataset.docs_iter():
    print(doc)`

it does not show anything. 

When I try
`import ir_datasets
dataset = ir_datasets.load('trec-robust04/fold1')
for doc in dataset.docs_iter()[:10]:
    print(doc)`

it just shows

> [INFO] [starting] building docstore
> docs_iter: 0it [00:00, ?it/s]
> [INFO] [finished] docs_iter: [00:00] [0it] [?it/s]
> [INFO] [finished] building docstore [40ms]

So, I can still not solve it. It will be very nice of you to provide more information about it. Thanks.

By the way, I will try the other solution you provide to  work it out.

**Comment by seanmacavaney** [comment]
@yuhaia have you tried this with the version on the master branch? We recently fixed an issue caused by a more recent release of the dataset (the data files themselves are not compressed anymore) and added better error messages.

**Comment by yuhaia** [comment]
> @yuhaia have you tried this with the version on the master branch? We recently fixed an issue caused by a more recent release of the dataset (the data files themselves are not compressed anymore) and added better error messages.

you mean I need to re-install ir_datasets package? By the way, I have tried this [irds_robust_anserini](https://github.com/seanmacavaney/irds_robust_anserini), and follow readme's `import ir_datasets
import irds_robust_anserini

dataset = ir_datasets.load('macavaney:anserini-trec-robust04')
`
then, when I run `for query in dataset.queries_iter():` it works. But for `for doc in dataset.docs_iter()[:2]:
    print(doc)` it still shows something like 

> [INFO] [starting] building docstore
docs_iter: 0it [00:00, ?it/s]
[INFO] [finished] docs_iter: [00:00] [0it] [?it/s]
[INFO] [finished] building docstore [66ms]

Did there exist further instructions to get the actual doc?
>

**Comment by seanmacavaney** [comment]
Yeah, I have not yet pushed those changes out to pypi. You can install the version from master like this:

```bash
pip install --upgrade git+https://github.com/allenai/ir_datasets.git
```

---

Regarding irds_robust_anserini-- it works for me. Here's what I get:

```python
import irds_robust_anserini
import ir_datasets
dataset = ir_datasets.load('macavaney:anserini-trec-robust04')
dataset.docs_iter()[:2]
```

> [INFO] [starting] building docstore
> [INFO] [starting] https://git.uwaterloo.ca/jimmylin/anserini-indexes/raw/master/index-robust04-20191213.tar.gz
> [INFO] [finished] https://git.uwaterloo.ca/jimmylin/anserini-indexes/raw/master/index-robust04-20191213.tar.gz: [00:17] [1.82GB] [104MB/s]
> [INFO] [starting] loading from lucene index                                                                                              
> [INFO] [finished] loading from lucene index: [10:58] [528030it] [802.12it/s]                                             
> docs_iter: 528030it [12:42, 692.26it/s]                
> [INFO] [finished] docs_iter: [12:42] [528030it] [692.26it/s]
> [INFO] [finished] building docstore [12:43]


From your output, it seems like it didn't even perform the download? Weird... Maybe there's something platform-dependent going wrong. What version of python are you using (I just tested on python 3.6 and 3.9)? Which operating system (just tested on Ubuntu 18.04 and 20.04)? Can you check what you have in `~/.ir_datasets/macavaney:anserini-trec-robust04` using:

```bash
du -hs ~/.ir_datasets/macavaney\:anserini-trec-robust04/*
```

> 1.2G	/home/sean/.ir_datasets/macavaney:anserini-trec-robust04/docs.pklz4
> 1016M	/home/sean/.ir_datasets/macavaney:anserini-trec-robust04/lucene_source.fdt

**Comment by yuhaia** [comment]
> Yeah, I have not yet pushed those changes out to pypi. You can install the version from master like this:
> 
> ```shell
> pip install --upgrade git+https://github.com/allenai/ir_datasets.git
> ```
> 
> Regarding irds_robust_anserini-- it works for me. Here's what I get:
> 
> ```python
> import irds_robust_anserini
> import ir_datasets
> dataset = ir_datasets.load('macavaney:anserini-trec-robust04')
> dataset.docs_iter()[:2]
> ```
> 
> > [INFO] [starting] building docstore
> > [INFO] [starting] https://git.uwaterloo.ca/jimmylin/anserini-indexes/raw/master/index-robust04-20191213.tar.gz
> > [INFO] [finished] https://git.uwaterloo.ca/jimmylin/anserini-indexes/raw/master/index-robust04-20191213.tar.gz: [00:17] [1.82GB] [104MB/s]
> > [INFO] [starting] loading from lucene index
> > [INFO] [finished] loading from lucene index: [10:58] [528030it] [802.12it/s]
> > docs_iter: 528030it [12:42, 692.26it/s]
> > [INFO] [finished] docs_iter: [12:42] [528030it] [692.26it/s]
> > [INFO] [finished] building docstore [12:43]
> 
> From your output, it seems like it didn't even perform the download? Weird... Maybe there's something platform-dependent going wrong. What version of python are you using (I just tested on python 3.6 and 3.9)? Which operating system (just tested on Ubuntu 18.04 and 20.04)? Can you check what you have in `~/.ir_datasets/macavaney:anserini-trec-robust04` using:
> 
> ```shell
> du -hs ~/.ir_datasets/macavaney\:anserini-trec-robust04/*
> ```
> 
> > 1.2G	/home/sean/.ir_datasets/macavaney:anserini-trec-robust04/docs.pklz4
> > 1016M	/home/sean/.ir_datasets/macavaney:anserini-trec-robust04/lucene_source.fdt


I have installed the version from master but still stay in the same situation.

As for the irds_robust_anserini, I run the code again and it's downloading the dataset. The version of python is 3.8 and my OS is MacOS Catalina 10.15.7. There may be some network error before. Anyway, I will check the result when the downloading process is finished. Thanks for your patient answers~

**Comment by seanmacavaney** [comment]
RE: irds_robust_anserini: that's very strange because it should throw an exception when a download fails due to network problems. (In fact, if the download doesn't yield the pre-defined hash for any reason, it should fail.) I'll write up an issue for this.

RE: master version: can you try removing everything in `~/.ir_datasets/trec-robust04` and starting from scratch? (Be sure to keep a copy of your corpus, though!)

**Comment by yuhaia** [comment]
> RE: irds_robust_anserini: that's very strange because it should throw an exception when a download fails due to network problems. (In fact, if the download doesn't yield the pre-defined hash for any reason, it should fail.) I'll write up an issue for this.
> 
> RE: master version: can you try removing everything in `~/.ir_datasets/trec-robust04` and starting from scratch? (Be sure to keep a copy of your corpus, though!)

I have removed everything in `~/.ir_datasets/trec-robust04` but when I started from scratch it shows the error like this `FileNotFoundError: [Errno 2] No such file or directory: '/Users/yuhai09/.ir_datasets/trec-robust04/queries.gz'`.  Does it not download `queries.gz` automatically?

**Comment by seanmacavaney** [comment]
Yup, it should download the queries automatically. What's the full stack trace there?

**Comment by yuhaia** [comment]
> Yup, it should download the queries automatically. What's the full stack trace there?

when I run the  code 
`import ir_datasets
dataset = ir_datasets.load('trec-robust04')

for query in dataset.queries_iter():
    print(query)`

the full stack trace is as follows:

`FileNotFoundError                         Traceback (most recent call last)
<ipython-input-31-b8d381492bd6> in <module>
      2 dataset = ir_datasets.load('trec-robust04')
      3 
----> 4 for query in dataset.queries_iter():
      5     print(query)
      6 

/opt/anaconda3/lib/python3.8/site-packages/ir_datasets/formats/trec.py in queries_iter(self)
    189     def queries_iter(self):
    190         fields, reading = {}, None
--> 191         with self._queries_dlc.stream() as f:
    192             f = codecs.getreader(self._encoding or 'utf8')(f)
    193             for line in f:

/opt/anaconda3/lib/python3.8/contextlib.py in __enter__(self)
    111         del self.args, self.kwds, self.func
    112         try:
--> 113             return next(self.gen)
    114         except StopIteration:
    115             raise RuntimeError("generator didn't yield") from None

/opt/anaconda3/lib/python3.8/site-packages/ir_datasets/util/fileio.py in stream(self)
    161     @contextlib.contextmanager
    162     def stream(self):
--> 163         with self._streamer.stream() as stream:
    164             yield gzip.GzipFile(fileobj=stream)
    165 

/opt/anaconda3/lib/python3.8/contextlib.py in __enter__(self)
    111         del self.args, self.kwds, self.func
    112         try:
--> 113             return next(self.gen)
    114         except StopIteration:
    115             raise RuntimeError("generator didn't yield") from None

/opt/anaconda3/lib/python3.8/site-packages/ir_datasets/util/download.py in stream(self)
    209                 yield stream
    210         else:
--> 211             with open(self.path(), 'rb') as f:
    212                 yield f
    213 

FileNotFoundError: [Errno 2] No such file or directory: '/Users/yuhai09/.ir_datasets/trec-robust04/queries.gz'
`

**Comment by seanmacavaney** [comment]
It's not obvious to me how it could get into this state because if the file doesn't exist, it downloads it. The code amounts to:

```
if already_found:
    return path
if path exists:
    alrady_found = True
    return path
else:
    download to path
    alrady_found = True
    return path
```

The one thing would be this wouldn't work if the python process is still running from before the files were cleared? It probably wouldn't hurt to add a check each time if the file exists...

**Comment by yuhaia** [comment]
> It's not obvious to me how it could get into this state because if the file doesn't exist, it downloads it. The code amounts to:
> 
> ```
> if already_found:
>     return path
> if path exists:
>     alrady_found = True
>     return path
> else:
>     download to path
>     already_found = True
>     return path
> ```
> 
> The one thing would be this wouldn't work if the python process is still running from before the files were cleared? It probably wouldn't hurt to add a check each time if the file exists...

I tried ir_datasets this morning again, and the `queries.gz` is downloaded. But When I follow the instruction to put `FBIS``FR94``FT`and`LATIMES` in  ~/.ir_datasets/trec-robust04/trec45/NEWS_data/, it stills shows:
`[INFO] [starting] building docstore
docs_iter: 0it [00:00, ?it/s]
[INFO] [finished] docs_iter: [00:00] [0it] [?it/s]
[INFO] [finished] building docstore [43ms]`

The good news is that after I run with irds_robust_anserini, I can see the documents! And I found it difficult to access the [trec website](https://trec.nist.gov/) sometimes, so the downloading issue maybe origins from that.

**Comment by seanmacavaney** [comment]
Thanks for all your help on this! I'm interested in getting to the bottom of these issues.

I'm aware of 2 formats that the documents can be distributed in. It seems like there may be a third. Can you help me out by running this so I can see what the contents of the source files you have look like?

```
du -a ~/.ir_datasets/trec-robust04/trec45/ | awk '{print $2}' | xargs md5sum
```

What's really strange is that it [passes the `expected_file_count` test](https://github.com/allenai/ir_datasets/blob/master/ir_datasets/datasets/trec_robust04.py#L36), but must just not find any documents in the files.

Regarding failing downloads from NIST-- the code is written to throw an exception when the contents of the downloaded file do not match a pre-defined hash. This works because automated tests [also periodically show problems downloading from NIST](https://github.com/allenai/ir_datasets/actions/runs/694637462), correctly caught as exceptions. I wrote up more about these strange failures on your system in #55 --- if you could add details there that may help to reproduce your issue, that'd be awesome!

**Comment by yuhaia** [comment]
> ```
> du -a ~/.ir_datasets/trec-robust04/trec45/ | awk '{print $2}' | xargs md5sum
> ```

Hi, when I run `du -a ~/.ir_datasets/trec-robust04/trec45/ | awk '{print $2}' | xargs md5sum` in my terminal, it shows that `xargs: md5sum: No such file or directory`.

**Comment by seanmacavaney** [comment]
Doesn't look like you have md5sum installed on your machine. How about just `du -a ~/.ir_datasets/trec-robust04/trec45/` then?

**Comment by yuhaia** [comment]
> Doesn't look like you have md5sum installed on your machine. How about just `du -a ~/.ir_datasets/trec-robust04/trec45/` then?

I found it a little longer, so I put the content to [here](https://github.com/yuhaia/robust04_folder_content)

**Comment by seanmacavaney** [comment]
Alright- so that matches one of the formats I am familiar with. (Strangely, the file sizes are double, but maybe that's a difference with the macos version of `du`?)

Are you sure you're using the `master` version of ir_datasets? Because this seems like the behavior exhibited on the current pypi version. One way to check would be to:
```bash
python
>>> import ir_datasets
>>> ir_datasets.datasets.trec_robust04.__file__
'/Users/sean/local/miniconda3/lib/python3.7/site-packages/ir_datasets/datasets/trec_robust04.py'
>>> exit()
# Then cat the contents of the above file:
cat '/Users/sean/local/miniconda3/lib/python3.7/site-packages/ir_datasets/datasets/trec_robust04.py'
```

Importantly, there should be this line:

```python
collection = TrecDocs(dlc['docs'], path_globs=['**/FBIS/FB*', '**/FR94/??/FR*', '**/FT/*/FT*', '**/LATIMES/LA*'], namespace=NAME, lang='en', expected_file_count=2295)
```

with the critical parts being that the `path_globs` do not contain `.gz` and the inclusion of the `expected_file_count`.

**Comment by yuhaia** [comment]
> ```shell
> ir_datasets.datasets.trec_robust04.__file__
> ```

Oh, sorry. I checked the version of ir_datasets with the way you provided and found that I am not using the `master` version of ir_datasets. Although I thought I have updated the version. Maybe I use the wrong environment before. 

Now, I install the `master` version of ir_datasets and the output of `trec_robust04.py` is the same as what you said. And it's building docstore after I run `for doc in dataset.docs_iter():`. It seems like everything is being fine. I will close this issue if it works. Thanks a lot for your time!
